Cabal/Distribution/InstalledPackageInfo.hs

```diff
@@ -249,6 +252,26 @@ parseExposedModules :: Parse.ReadP r [ExposedModule]
 parseExposedModules = parseOptCommaList parse
 
 -- -----------------------------------------------------------------------------
+-- ABI dependency
+
+data AbiDependency = AbiDependency {
```

Comment:
We really ought to document these things. How about:

> This is here to help catch obvious wrongness. By computing and storing the `abiHash` of each package, and by annotating them on each dependency then we can do an extra sanity check that the ABIs match. This can catch cases where users or tools make mistakes, especially in combining packages from multiple databases.
> 
> It's worth noting however that just because an ABI matches, does not mean that the package is the one you expect. Many changes cannot be caught as ABI mismatches. So if you're a tool author and see these mismatches happening then you should think carefully about your scheme, especially how the unit id is chosen.